File: packages/cli/src/commands/report.ts
import { Command } from 'commander';
import { createInterface } from 'node:readline/promises';
import { icons, c, log, logJson, logError, withSpinner, getCurrentMode } from '../lib/output.js';
import { loadConfig } from '../lib/config.js';
import { createApiClient, ApiClientError } from '../lib/api-client.js';

/**
 * Prompt the user for a reason string.
 */
const promptReason = async (): Promise<string | null> => {
  const rl = createInterface({ input: process.stdin, output: process.stdout });
  try {
    const answer = await rl.question('  Reason for report: ');
    const trimmed = answer.trim();
    return trimmed.length > 0 ? trimmed : null;
  } finally {
    rl.close();
  }
};

export const registerReportCommand = (program: Command): void => {
  program
    .command('report <name>')
    .description('Report a skill for policy violations')
    .option('--reason <reason>', 'Reason for the report')
    .option('--detail <detail>', 'Additional detail')
    .action(async (name: string, opts: { reason?: string; detail?: string }) => {
      const mode = getCurrentMode();

      // 1. Check auth
      const config = loadConfig();
      if (!config.token) {
        logError(
          'Not authenticated',
          'You must be logged in to report skills.',
          `Run ${c.cmd('spm login')} to authenticate.`,
        );
        process.exitCode = 1;
        return;
      }

      // 2. Get reason (from flag or prompt)
      let reason = opts.reason;
      if (!reason) {
        reason = (await promptReason()) ?? undefined;
        if (!reason) {
          log(`${icons.info} Report cancelled.`);
          return;
        }
      }

      const api = createApiClient(config);

      // 3. Submit report
      try {
        await withSpinner(`Submitting report for ${name}...`, () =>
          api.reportSkill(name, { reason, detail: opts.detail }),
        );

        log(`${icons.success} Report submitted for ${c.name(name)}`);

        if (mode === 'json') {
          logJson({
            command: 'report',
            status: 'success',
            name,
            reason,
            detail: opts.detail,
          });
        }
      } catch (err) {
        if (err instanceof ApiClientError) {
          if (err.status === 404) {
            logError('Skill not found', `${name} does not exist in the registry.`);
          } else if (err.status === 401) {
            logError(
              'Authentication failed',
              'Your token may have expired.',
              `Run ${c.cmd('spm login')} to re-authenticate.`,
            );
          } else if (err.status === 429) {
            logError('Rate limited', 'Too many reports submitted. Please try again later.');
          } else {
            logError('Report failed', err.apiError.message, err.apiError.suggestion);
          }
        } else {
          logError('Report failed', err instanceof Error ? err.message : String(err));
        }
        process.exitCode = 1;
      }
    });
};

File: cli-go/cmd/report.go
package cmd

import (
	"fmt"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var (
	reportReason   string
	reportPriority string
)

var reportCmd = &cobra.Command{
	Use:   "report <name>",
	Short: "Report a skill for policy violations",
	Long:  "Submit a report for a skill that violates registry policies.",
	Args:  cobra.ExactArgs(1),
	RunE:  runReport,
}

func init() {
	reportCmd.Flags().StringVar(&reportReason, "reason", "", "Reason for reporting (required)")
	reportCmd.Flags().StringVar(&reportPriority, "priority", "normal", "Report priority (low, normal, high)")
	_ = reportCmd.MarkFlagRequired("reason")
	rootCmd.AddCommand(reportCmd)
}

func runReport(_ *cobra.Command, args []string) error {
	name := args[0]

	// 1. Check auth
	if Cfg.Token == "" {
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(map[string]any{
				"command": "report",
				"status":  "error",
				"error":   "not_logged_in",
			})
		}
		Out.LogError("Not authenticated. Run %s to log in.", output.Cyan("spm login"))
		return fmt.Errorf("not logged in")
	}

	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)

	sp := Out.StartSpinner(fmt.Sprintf("Submitting report for %s...", name))
	resp, err := client.Report(name, reportReason, reportPriority)
	output.StopSpinner(sp)

	if err != nil {
		var apiErr *api.APIError
		if isAPIError(err, &apiErr) {
			if Out.Mode == output.ModeJSON {
				return Out.LogJSON(map[string]any{
					"command": "report",
					"status":  "error",
					"error":   apiErr.Code,
					"message": apiErr.Message,
				})
			}
			if apiErr.IsNotFound() {
				Out.LogError("Skill %s does not exist in the registry.", name)
			} else if apiErr.IsUnauthorized() {
				Out.LogError("Authentication failed. Run %s to re-authenticate.", output.Cyan("spm login"))
			} else if apiErr.IsRateLimited() {
				Out.LogError("Rate limited. Too many reports submitted. Please try again later.")
			} else {
				Out.LogError("Report failed: %s", apiErr.Message)
			}
			return fmt.Errorf("report failed: %s", apiErr.Message)
		}
		Out.LogError("Report failed: %s", err)
		return fmt.Errorf("report failed: %w", err)
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(map[string]any{
			"command":    "report",
			"status":     "success",
			"id":         resp.ID,
			"skill":      resp.Skill,
			"reason":     reportReason,
			"priority":   reportPriority,
			"created_at": resp.CreatedAt,
		})
	}

	Out.Log("%s Report submitted for %s", output.Green(output.Icons["success"]), output.Cyan(name))
	Out.Log("  Report ID: %s", output.Dim(resp.ID))

	return nil
}
